renderdoc/driver/vulkan/vk_pixelhistory.cpp

```diff
@@ -0,0 +1,604 @@
+#include <stdio.h>
+#include <sys/types.h>
+#include <vector>
+#include "driver/vulkan/vk_debug.h"
+#include "driver/vulkan/vk_replay.h"
+#include "vk_shader_cache.h"
+
+bool isUsageUAV(ResourceUsage usage)
+{
+  return ((usage >= ResourceUsage::VS_RWResource && usage <= ResourceUsage::CS_RWResource) ||
+          usage == ResourceUsage::CopyDst || usage == ResourceUsage::Copy ||
+          usage == ResourceUsage::Resolve || usage == ResourceUsage::ResolveDst ||
+          usage == ResourceUsage::GenMips);
+}
+
+struct CopyPixelParams
+{
+  bool multisampled;
+  bool floatTex;
+  bool uintTex;
+  bool intTex;
+
+  bool depthcopy;
+  VkImage srcImage;
+  VkFormat srcImageFormat;
+  VkImageLayout srcImageLayout;
+  VkOffset3D imageOffset;
+
+  VkBuffer dstBuffer;
+};
+
+struct PixelHistoryResources
+{
+  VkBuffer dstBuffer;
+  VkDeviceMemory bufferMemory;
+
+  // Used for offscreen rendering for draw call events.
+  VkImage colorImage;
+  VkImageView colorImageView;
+  VkImage stencilImage;
+  VkImageView stencilImageView;
+  VkDeviceMemory gpuMem;
+};
+
+struct PixelHistoryValue
+{
+  uint8_t color[16];
+  uint8_t depth[8];
+};
+
+struct EventInfo
+{
+  PixelHistoryValue premod;
+  PixelHistoryValue postmod;
+  uint8_t shadout[16];
+  uint8_t fixed_stencil[8];
+  uint8_t original_stencil[8];
+};
+
+struct VulkanPixelHistoryCallback : public VulkanDrawcallCallback
+{
+  VulkanPixelHistoryCallback(WrappedVulkan *vk, uint32_t x, uint32_t y, VkImage image,
+                             VkFormat format, uint32_t sampleMask, VkQueryPool occlusionPool,
+                             VkImageView colorImageView, VkImageView stencilImageView,
+                             VkImage colorImage, VkImage stencilImage, VkBuffer dstBuffer,
+                             const std::vector<EventUsage> &events)
+      : m_pDriver(vk),
+        m_X(x),
+        m_Y(y),
+        m_Image(image),
+        m_Format(format),
+        m_DstBuffer(dstBuffer),
+        m_SampleMask(sampleMask),
+        m_OcclusionPool(occlusionPool),
+        m_ColorImageView(colorImageView),
+        m_StencilImageView(stencilImageView),
+        m_ColorImage(colorImage),
+        m_StencilImage(stencilImage),
+        m_PrevState(vk, NULL)
+  {
+    m_pDriver->SetDrawcallCB(this);
+    for(size_t i = 0; i < events.size(); i++)
+      m_Events.insert(std::make_pair(events[i].eventId, events[i]));
+  }
+  ~VulkanPixelHistoryCallback() { m_pDriver->SetDrawcallCB(NULL); }
+  void CopyPixel(uint32_t eid, VkCommandBuffer cmd, size_t offset)
+  {
+    CopyPixelParams colourCopyParams = {};
+    colourCopyParams.multisampled = false;    // TODO: multisampled
+    colourCopyParams.srcImage = m_Image;
+    colourCopyParams.srcImageLayout =
+        VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;    // TODO: image layout
+    colourCopyParams.srcImageFormat = m_Format;
+    colourCopyParams.imageOffset = VkOffset3D{int32_t(m_X), int32_t(m_Y), 0};
```

Comment:
Prefer not using this modern C++ style with brace constructors. Constructing from a plain list of items is fine but if you invoke a constructor it should be with normal brackets.

Reply:
DOne